# Segmenting & Clustering Neighborhoods in Toronto (Part 3)

This notebook was prepared as the part of Data Science Capstone project assignment in Coursersa. This notebook will cover how to explore, segment, and cluster the neighborhoods in the city of Toronto by using the neighborhoods data from Wikipedia.

## Explore the Data
### Part 1 - Scarpe Wikipedia for Data
To get the neighborhoods data for this project, we will use this page from wikipedia: [List of postal codes of Canada: M](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M).

On the wikipedia page we will have the list of postal code of Canada which started with "M", we will use the table as the source of our data in this project. In order to get data ready for the analysis we need to scarpe the website and convert the table into _Pandas_ dataframe. To do so, we will utilize _BeuatifulSoup_ library which will allow us to scrape the wikipedia page and parse the table to be used in our analysis.

Before we start with our code we could inspect the website to get to know what we should be looking for in the website code. For this purpose we could use _Chrome_ developer option to _inspect_ the element of the website and find our table.

![wiki_inspect_code](img/wiki_inspect.png)

As we can see from the image above, the header and the row of the table can be easily identified from the source code. We also able to identify which class contain the table that we're looking for, in this case table class `"wikitable sortable jquery-tablesorter"`. We will use these information for our next step when parsing the table into _Pandas_ dataframe.


#### Utilizing _BeautifulSoup_
First we're going to use _requests_ library which will allow us to send _organic, grass-fed HTTP/1.1_ requests.

In [2]:
import requests
#ping a website and return HTML of the website.
url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

from bs4 import BeautifulSoup
soup = BeautifulSoup(url,'lxml')
#find the table class "wikitable sortable"
dtable = soup.find('table',{'class':'wikitable sortable'})

Next we would like to do some iteration on the row and append the value into a list which will then converted into pandas dataframe.

In [3]:
import pandas as pd

drow = dtable.findChildren(['th', 'tr'])

#iterate the row and append the value into a list called tes
tes=[]
for row in drow:
    tes1=[]
    cells = row.findChildren('td')
    for cell in cells:
        value = cell.string
        value = value.replace('\n','')
        value = value.replace(' /', ',')
        tes1.append(value)
    tes.append(tes1)

df = pd.DataFrame(tes)
df = df.rename(columns={0: 'PostalCode', 1: 'Borough', 2: 'Neighborhood'})
df.head(10)

PostalCode           Borough                      Neighborhood
0       None              None                              None
1       None              None                              None
2       None              None                              None
3       None              None                              None
4        M1A      Not assigned                                  
5        M2A      Not assigned                                  
6        M3A        North York                         Parkwoods
7        M4A        North York                  Victoria Village
8        M5A  Downtown Toronto         Regent Park, Harbourfront
9        M6A        North York  Lawrence Manor, Lawrence Heights

Let's do some clean up on the dataframe, so it follow these requirements:
* The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
* Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
* More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma.
* If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [4]:
import numpy as np

#clean up dataframe for null value in Postal Code
null_mask = df['PostalCode'].notnull().values
df = df[null_mask]

#clean up dataframe for 'Not assigned' Borough
borough_mask = df['Borough'].values != 'Not assigned'
df = df[borough_mask]

#put value into 'Not assigned' Neighborhood
df['Neighborhood'] = np.where(df['Neighborhood'] == 'Not assigned', df['Borough'], df['Neighborhood'])

#reset index after clean up
df.set_index(['PostalCode'], inplace=True)
df.reset_index(level=['PostalCode'], inplace=True)
df.head(10)

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5        M9A         Etobicoke                             Islington Avenue
6        M1B       Scarborough                               Malvern, Rouge
7        M3B        North York                                    Don Mills
8        M4B         East York              Parkview Hill, Woodbine Gardens
9        M5B  Downtown Toronto                     Garden District, Ryerson

Now, let's check how many unique values available in the `PostalCode` column and then compare it to the size of the dataframe. If we have a good clean up we should end up with the same number of `PostalCode` unique value vs number of rows in the dataframe.

In [5]:
print (len(df['PostalCode'].unique()))

103


From the above code, we've got 103 unique values in our `PostalCode` column. Now, let's take a look on the number of rows we had in our dataframe. We should get the same number if we had our clean up right.

In [6]:
print (df.shape)

(103, 3)


### Part 2 - Assigning Geo-Location

In order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood. This step can be done automatically by utilizing Geocoder or the provided _.csv_ file, do the the unreliability of the package I choose to use the provided _.csv_ file in our case here.

We will load the geospatial _.csv_ file as a new _Pandas_ dataframe and combine the information with our previously created dataframe.

In [7]:
df_geo = pd.read_csv('Geospatial_Coordinates.csv')
df_geo.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

Now, let's combine the latitude and longitude information from `df_geo` to `df_geo` by merging the two dataframe on `PostalCode` value and drop the duplicates column.

In [8]:
df_fin = df.merge(df_geo, on='PostalCode', how='left').drop_duplicates()
df_fin.head(10)

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5        M9A         Etobicoke                             Islington Avenue   
6        M1B       Scarborough                               Malvern, Rouge   
7        M3B        North York                                    Don Mills   
8        M4B         East York              Parkview Hill, Woodbine Gardens   
9        M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

### Part-3 Explore and Cluster the Neighborhood

On this last part, we will explore and cluster the neighborhoods in Toronto, before we do that let's import all the required library.

In [9]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


#### Use geopy library to get the latitude and longitude values of Toronto
In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent my_explorer, as shown below.

In [11]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="my_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [13]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_fin['Latitude'], df_fin['Longitude'], df_fin['Borough'], df_fin['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

For illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods with Borough containt Toronto. So let's slice the original dataframe and create a new dataframe of the Toronto data.

In [20]:
toronto_data = df_fin[df_fin['Borough'].str.contains('Toronto')].reset_index(drop=True)
toronto_data.head()

PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

Let's visualize the neighborhoods in a map.

In [21]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [87]:
# @hidden_cell
CLIENT_ID = '*****' # your Foursquare ID
CLIENT_SECRET = '*****' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

#### Let's create a function to get venues from Foursquare for all neighborhoods in Toronto

In [27]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now write the code to run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [28]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )


Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town,

In [29]:
print(toronto_venues.shape)
toronto_venues.head()

(1623, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

Let's check how many venues were returned for each neighborhood

In [30]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre                                  18   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     63   
Christie                                                               17   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           19   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                5   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          23   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              55   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               41   
Moore Park, Summerhill East                                             1   
North Toronto West                                                     20   
Parkdale, Roncesvalles                                                 13   
Queen's Park, Ontario Provincial Government                            38   
Regent Park, Harbourfront                                              48   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     39   
St. James Town                                                         77   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         95   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     17   
The Annex, North Midtown, Yorkville                                    22   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre                                   18   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      63   
Christie                                                                17   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              36   
Davisville North                    

Let's find out how many unique categories can be curated from all the returned venues

In [31]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 231 uniques categories.


### Part-4 Analyze Each Neighborhoods

In [38]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
toronto_onehot.set_index(['Neighborhood'], inplace=True)
toronto_onehot.reset_index(level=['Neighborhood'], inplace=True)
print (toronto_onehot.shape)
toronto_onehot.head()

(1623, 231)


Neighborhood  Airport  Airport Food Court  Airport Gate  \
0  Regent Park, Harbourfront        0                   0             0   
1  Regent Park, Harbourfront        0                   0             0   
2  Regent Park, Harbourfront        0                   0             0   
3  Regent Park, Harbourfront        0                   0             0   
4  Regent Park, Harbourfront        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       1     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               1   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Candy Store  \
0        0     0                          0            0   
1        0     0                          0            0   
2        0     0                          0            0   
3        0     0                          0     

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [39]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood   Airport  \
0                                         Berczy Park  0.000000   
1        Brockton, Parkdale Village, Exhibition Place  0.000000   
2               Business reply mail Processing Centre  0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...  0.058824   
4                                  Central Bay Street  0.000000   
5                                            Christie  0.000000   
6                                Church and Wellesley  0.000000   
7                      Commerce Court, Victoria Hotel  0.000000   
8                                          Davisville  0.000000   
9                                    Davisville North  0.000000   
10                       Dufferin, Dovercourt Village  0.000000   
11             First Canadian Place, Underground city  0.000000   
12                           Forest Hill North & West  0.000000   
13                           Garden District, Ryerson  0.000000   
14  Harbourfront East, Union Station, Toronto Islands  0.000000   
15                      High Park, The Junction South  0.000000   
16                     India Bazaar, The Beaches West  0.000000   
17          Kensington Market, Chinatown, Grange Park  0.000000   
18                                      Lawrence Park  0.000000   
19                           Little Portugal, Trinity  0.000000   
20                        Moore Park, Summerhill East  0.000000   
21                                 North Toronto West  0.000000   
22                             Parkdale, Roncesvalles  0.000000   
23        Queen's Park, Ontario Provincial Government  0.000000   
24                          Regent Park, Harbourfront  0.000000   
25                           Richmond, Adelaide, King  0.000000   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                 Runnymede, Swansea  0.000000   
29                                     St. James Town  0.000000   
30                        St. James Town, Cabbagetown  0.000000   
31                                     Stn A PO Boxes  0.000000   
32                                    Studio District  0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...  0.000000   
34                The Annex, North Midtown, Yorkville  0.000000   
35                                        The Beaches  0.000000   
36                       The Danforth West, Riverdale  0.000000   
37           Toronto Dominion Centre, Design Exchange  0.000000   
38                     University of Toronto, Harbord  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.000000      0.000000        0.000000         0.000000   
3             0.058824      0.058824        0.117647         0.176471   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000        0.000000         0.000000   
11            0.000000      0.000000        0.000000         0.000000   
12            0.000000      0.000000        0.000000         0.000000   
13            0.000000      0.000000        0.000000         0.000000   
14            0.000000      0.000000        0.000000         0.000000   
15            0.000000      0.000000        0.000000         0.000000   
16            0.000000      0.000000        0.000000         0.000000   
17            0.000000      0.000000        0.

Let's confirm the new dataframe size.

In [42]:
toronto_grouped.shape

(39, 231)

Let's print each neighborhood along with the top 5 most common venues

In [43]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2          Restaurant  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1  Breakfast Spot  0.09
2       Nightclub  0.09
3     Coffee Shop  0.09
4             Gym  0.04


----Business reply mail Processing Centre----
                  venue  freq
0           Yoga Studio  0.06
1                   Spa  0.06
2  Gym / Fitness Center  0.06
3         Garden Center  0.06
4                Garden  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3           Airport  0.06
4               Bar  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.17
1                 Café  0

#### Let's put that into a pandas dataframe
First, let's write a function to sort the venues in descending order.

In [44]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [77]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2              Business reply mail Processing Centre           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar    Seafood Restaurant                  Café   
1        Breakfast Spot           Coffee Shop             Nightclub   
2         Auto Workshop    Light Rail Station            Skate Park   
3        Airport Lounge      Airport Terminal               Airport   
4    Italian Restaurant                  Café        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery           Cheese Shop            Restaurant   
1         Burrito Place            Restaurant    Italian Restaurant   
2            Smoke Shop         Burrito Place        Farmers Market   
3       Harbor / Marina                 Plane   Rental Car Location   
4           Salad Place   Fried Chicken Joint       Bubble Tea Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Beer Bar                   Pub  Portuguese Restaurant  
1          Intersection               Stadium                    Bar  
2                   Spa  Fast Food Restaurant             Restaurant  
3      Sculpture Garden         Boat or Ferry                    Bar  
4          Burger Joint        Ice Cream Shop                    Bar

### Part-5 Cluster the Neighborhood

Run k-means to cluster the neighborhood into 5 clusters.

In [78]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [79]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               0     Health Food Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub                Bakery                  Park   
1      Sushi Restaurant                 Diner    Mexican Restaurant   
2           Coffee Shop                  Café            Restaurant   
3                  Café             Gastropub   American Restaurant   
4                 Trail                   Pub           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0        Breakfast Spot                  Café                    Theater   
1          Burger Joint                  Bank                        Bar   
2    Italian Restaurant       Bubble Tea Shop  Middle Eastern Restaurant   
3          Cocktail Bar        Cosmetics Shop           Department Store   
4         Deli / Bodega  Ethiopian Restaurant          Electronics Store   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Yoga Studio    Mexican Restaurant             Shoe Store  
1               Discount Store           Yoga Studio    Fried Chicken Joint  
2               Cosmetics Shop   Japanese Restaurant              Bookstore  
3               Lingerie Store   Moroccan Restaurant     Italian Restaurant  
4  Eastern European Restaurant            Donut Shop       Doner Restaurant

Finally, let's visualize the resulting clusters

In [80]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Part-6 Examine Clusters

Now, we can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories,we can then assign a name to each cluster.

#### Cluster 1

In [81]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0        Clothing Store   
3   Downtown Toronto               0           Coffee Shop   
4       East Toronto               0     Health Food Store   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0         Grocery Store   
8   Downtown Toronto               0           Coffee Shop   
9       West Toronto               0                Bakery   
10  Downtown Toronto               0           Coffee Shop   
11      West Toronto               0                   Bar   
12      East Toronto               0      Greek Restaurant   
13  Downtown Toronto               0           Coffee Shop   
14      West Toronto               0                  Café   
15      East Toronto               0        Sandwich Place   
16  Downtown Toronto               0           Coffee Shop   
17      East Toronto               0                  Café   
20   Central Toronto               0      Department Store   
22      West Toronto               0       Thai Restaurant   
23   Central Toronto               0        Clothing Store   
24   Central Toronto               0        Sandwich Place   
25      West Toronto               0             Gift Shop   
26   Central Toronto               0        Sandwich Place   
27  Downtown Toronto               0                  Café   
28      West Toronto               0           Coffee Shop   
30  Downtown Toronto               0                  Café   
31   Central Toronto               0           Coffee Shop   
32  Downtown Toronto               0       Airport Service   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0           Coffee Shop   
36  Downtown Toronto               0           Coffee Shop   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0           Yoga Studio   

   2nd Most Common Venue    3rd Most Common Venue  \
0                    Pub                   Bakery   
1       Sushi Restaurant                    Diner   
2            Coffee Shop                     Café   
3                   Café                Gastropub   
4                  Trail                      Pub   
5           Cocktail Bar       Seafood Restaurant   
6     Italian Restaurant                     Café   
7                   Café                     Park   
8                   Café               Restaurant   
9               Pharmacy            Grocery Store   
10              Aquarium                    Hotel   
11            Restaurant    Vietnamese Restaurant   
12           Coffee Shop       Italian Restaurant   
13                 Hotel                     Café   
14        Breakfast Spot              Coffee Shop   
15  Fast Food Restaurant                Pet Store   
16            Restaurant                     Café   
17           Coffee Shop                Gastropub   
20        Sandwich Place                     Park   
22    Mexican Restaurant                     Café   
23           Coffee Shop   Furniture / Home Store   
24                  Café              Coffee Shop   
25          Dessert Shop         Cuban Restaurant   
26           Pizza Place             Dessert Shop   
27                   Bar       Italian Restaurant   
28                  Café              Pizza Place   
30           Coffee Shop       Mexican Restaurant   
31                   Pub  Health & Beauty Service   
32        Airport Lounge         Airport Terminal   
34                  Café       Seafood Restaurant   
35    Italian Restaurant               Restaurant   
36                  Café                    Hotel   
37      Sushi Restaurant      Japanese Restaurant   
38         Auto Workshop       Light Rai

#### Cluster 2

In [82]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19           Yoga Studio      Department Store    Falafel Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19           Event Space  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19  Eastern European Restaurant            Donut Shop       Doner Restaurant

#### Cluster 3

In [83]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
29  Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29           Yoga Studio      Department Store           Event Space   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
29  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29            Donut Shop      Doner Restaurant                Dog Run

#### Cluster 4

In [84]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21   Central Toronto               3                  Park   
33  Downtown Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21         Jewelry Store                 Trail              Bus Line   
33            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21      Sushi Restaurant           Yoga Studio          Dessert Shop   
33         Deli / Bodega  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue  \
21         Ethiopian Restaurant     Electronics Store   
33  Eastern European Restaurant            Donut Shop   

         10th Most Common Venue  
21  Eastern European Restaurant  
33             Doner Restaurant

#### Cluster 5

In [85]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               4                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18  Construction & Landscaping              Bus Line           Swim School   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18           Yoga Studio          Dessert Shop           Event Space   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
18  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

### Part-7 Cluster Analysis
After examining the cluster result, we could see do some analysis on the result and possibly get some idea on what is happening on each clusters. From the above cluster we could summarize the following:

* Cluster-1, a downtown area with the most common venue are coffee shop and cafe
* Cluster-2, close to the residential area with the most common venue being a garden
* Cluster-3, a recreational area with the most common venue being a park
* Cluster-4, an outdoor recreational area with couple of Trail for outdoor hiking
* Cluster-5, Park is still the most common area but there is a school and a bus line which possibly a school district
